lib/Monads/NonDetMonadVCG.thy

```diff
@@ -1476,123 +1135,106 @@ lemma hoare_drop_impE_R:
   by (auto simp: validE_E_def validE_def valid_def split: sum.splits)
 
-(*
- * \<lbrakk>\<And>x. \<lbrace>?B1 x\<rbrace> ?g1 x -, \<lbrace>?E\<rbrace>; \<lbrace>?P\<rbrace> ?f1 \<lbrace>?B1\<rbrace>, \<lbrace>?E\<rbrace>\<rbrakk> \<Longrightarrow> \<lbrace>?P\<rbrace> ?f1 >>=E ?g1 -, \<lbrace>?E\<rbrace>
- *)
-lemmas [wp_split] =
-  validE_validE_E [OF hoare_vcg_seqE [OF validE_E_validE]]
+lemma hoare_vcg_set_pred_lift:
+  assumes "\<And>P x. m \<lbrace> \<lambda>s. P (f x s) \<rbrace>"
+  shows "m \<lbrace> \<lambda>s. P {x. f x s} \<rbrace>"
+  using assms[where P="\<lambda>x . x"] assms[where P=Not] use_valid
+  by (fastforce simp: valid_def elim!: subst[rotated, where P=P])
 
-lemma case_option_wp:
-  assumes x: "\<And>x. \<lbrace>P x\<rbrace> m x \<lbrace>Q\<rbrace>"
-  assumes y: "\<lbrace>P'\<rbrace> m' \<lbrace>Q\<rbrace>"
-  shows      "\<lbrace>\<lambda>s. (x = None \<longrightarrow> P' s) \<and> (x \<noteq> None \<longrightarrow> P (the x) s)\<rbrace>
-                case_option m' m x \<lbrace>Q\<rbrace>"
-  apply (cases x; simp)
-   apply (rule y)
-  apply (rule x)
-  done
+lemma hoare_vcg_set_pred_lift_mono:
+  assumes f: "\<And>x. m \<lbrace> f x \<rbrace>"
+  assumes mono: "\<And>A B. A \<subseteq> B \<Longrightarrow> P A \<Longrightarrow> P B"
+  shows "m \<lbrace> \<lambda>s. P {x. f x s} \<rbrace>"
+  by (fastforce simp: valid_def elim!: mono[rotated] dest: use_valid[OF _ f])
 
-lemma case_option_wpE:
-  assumes x: "\<And>x. \<lbrace>P x\<rbrace> m x \<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
-  assumes y: "\<lbrace>P'\<rbrace> m' \<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
-  shows      "\<lbrace>\<lambda>s. (x = None \<longrightarrow> P' s) \<and> (x \<noteq> None \<longrightarrow> P (the x) s)\<rbrace>
-                case_option m' m x \<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
-  apply (cases x; simp)
-   apply (rule y)
-  apply (rule x)
-  done
 
-lemma in_bindE:
-  "(rv, s') \<in> fst ((f >>=E (\<lambda>rv'. g rv')) s) =
-  ((\<exists>ex. rv = Inl ex \<and> (Inl ex, s') \<in> fst (f s)) \<or>
-  (\<exists>rv' s''. (rv, s') \<in> fst (g rv' s'') \<and> (Inr rv', s'') \<in> fst (f s)))"
-  apply (rule iffI)
-   apply (clarsimp simp: bindE_def bind_def)
-   apply (case_tac a)
-    apply (clarsimp simp: lift_def throwError_def return_def)
-   apply (clarsimp simp: lift_def)
-  apply safe
-   apply (clarsimp simp: bindE_def bind_def)
-   apply (erule rev_bexI)
-   apply (simp add: lift_def throwError_def return_def)
-  apply (clarsimp simp: bindE_def bind_def)
-  apply (erule rev_bexI)
-  apply (simp add: lift_def)
-  done
+subsection \<open>Strongest postcondition rules\<close>
 
-(*
- * \<lbrace>?P\<rbrace> ?m1 -, \<lbrace>?E\<rbrace> \<Longrightarrow> \<lbrace>?P\<rbrace> liftME ?f1 ?m1 -, \<lbrace>?E\<rbrace>
- *)
-lemmas [wp_split] = validE_validE_E [OF liftME_wp, simplified, OF validE_E_validE]
+lemma get_sp:
+  "\<lbrace>P\<rbrace> get \<lbrace>\<lambda>a s. s = a \<and> P s\<rbrace>"
```

Comment:
Are these strongest postcondition rules a little easier to read if the return value has name `rv` or something?

Reply:
Yes, I should make that part of the style update.